src/metabase/async/streaming_response.clj

```diff
@@ -57,8 +58,15 @@
     (with-open [os os]
       (log/trace (u/pprint-to-str (list 'write-error! obj)))
       (try
-        (with-open [writer (BufferedWriter. (OutputStreamWriter. os StandardCharsets/UTF_8))]
-          (json/generate-stream obj writer))
+        (let [obj (if (map? obj)
```

Comment:
A couple thoughts here:

- nit: you don't need the outer `if` because `walk/prewalk` will apply the fn to the whole `obj` first, so you can just do 

```
(let [obj (walk/prewalk (fn [x]
                                  (if (map? x)
                                    (apply dissoc x [:visualization_settings :visualization-settings :viz-settings :preprocessed :column_settings :column-settings])
                                    x))
                                  obj)]
          (with-open [writer (BufferedWriter. (OutputStreamWriter. os StandardCharsets/UTF_8))]
            (json/generate-stream obj writer)))
```

- does this whole thing belong in `format-exception` instead of `write-error!` maybe? It feels like trying to sanitize an arbitrary map is a hard problem with only hacky solutions, whereas if we're sanitizing a Throwable it seems like we could just remove all ex-data. Something like `(update (Throwable->map e) :via (fn [cause-chain] (map #(dissoc % :data) cause-chain)))` maybe?

Reply:
I'd say the 'hacky' part about this problem is that we're writing a stacktrace to a downloaded file at all. I think the reason we do this is to communicate *something* to the user that a download has failed, rather than giving nothing. 🤷🏼

Reply:
Sorry, "hacky" might be too harsh a way of putting it. I guess what I mean is that if `write-error!` is meant to handle an unbounded set of potential `obj` maps, it feels slightly awkward to then turn around and make assumptions about the content of those maps.

We know the current set of problematic keys, but what if someone throws a new exception somewhere that has newly problematic keys in `ex-data`? Unless there's a strong reason to give the user the ex-data, it feels simpler to just give the error messages and stacktrace and be reasonably sure we're not exposing something we shouldn't.

Reply:
@johnswanson, I've made a change and am curious what you think. Here's some detail:

In the case of errors being saved to files, the obj is a map with a :status :failed key assoc'd by the qp somewhere
along the way. The format-exception cond branch is not used in this case, so we can't rely on modifying
`format-exception` to just dissoc ex-data.

Instead, I've kept the obj modification, but it excludes the query and the preprocessed keys, which are the only 2
keys where we see :viz-settings. This doesn't eliminate the problem of needing to exclude other keys in the future,
but it does improve the surface area - the query and viz settings are the most likely candidates for holding
potentially sensitive info.

I'd prefer to stick with this solution, as we can get a fix for the security concern in quicker, and perhaps a
non-security bug can be opened regarding the most correct way to indicate errors to users whose downloads have failed.

Reply:
Alright, that makes sense to me. Thanks!